# Solving Systems of Linear Equations with NumPy - Lab

## Introduction 

Now we have gathered all the required skills needed to solve systems of linear equations. We saw why need to calculate inverses of matrices, followed by matrix multiplication to figure the values of unknown variables. 

Following exercises present some problems that can be converted into a system of linear equations. Solve the problems for the values of unknown variables. 

## Objectives
You will be able to:

* Build a system of equations from given problem by creating matrix and vectors in NumPy
* Calculate the inverse and dot-product to get the values for unknown variables.
* Solve linear equations using `numpy.linalg.solve()`

## Exercise 1

A coffee shop is having a sale on coffee and tea. 

On day 1, 29 bags of coffee and 41 bags of tea was sold, giving a total of 490 dollars.

On day 2, they sold 23 bags of coffee and 41 bags of tea, for which customers paid a total of 448 dollars.  

How much does each bag cost?

In [1]:
import numpy as np

In [8]:
# Solution:
A = np.matrix([[29,41], [23,41]])
B = np.matrix([490,448])
B = B.T
A_inv = np.linalg.inv(A)
X = A_inv.dot(B)
print(X)

[[7.]
 [7.]]


In [9]:
A.dot(X)

matrix([[490.],
        [448.]])

In [10]:
np.linalg.solve(A,B)

matrix([[7.],
        [7.]])

# Explain your answer in terms of prices

each bag of tea AND coffee cost $7

## Exercise 2:

The cost of admission to a popular music concert was 162 dollars for 12 children and 3 adults. 

The admission was 122 dollars for 8 children and 3 adults in another music concert. 

How much was the admission for each child and adult?

In [12]:
# Create a system of equations

A = np.matrix([[12,3], [8,3]])
B = np.matrix([162,122])
B = B.T
A_inv = np.linalg.inv(A)
X = A_inv.dot(B)
print(X.T)

[[10. 14.]]


In [13]:
A.dot(X)

matrix([[162.],
        [122.]])

In [14]:
np.linalg.solve(A,B)

matrix([[10.],
        [14.]])

# Describe your result
children's tickets are $10
adult's tickets are $14

## Exercise 3:

You want to make a certain kind of tropical punch, using bananas, oranges, and papayas. 

Suppose you don't know how many of each to put in the punch, but you know that there are seven pieces of fruit in the mix, and there are twice as many oranges as bananas. 

You also know that the seven pieces of fruit cost \$5.25, 

where bananas cost \$.50 each, oranges cost \$.75 each, and papayas cost \$1.25 each.

How many bananas, oranges and papayas would you need to make this punch ?

b + o + p = 7
2b = o
.5b + .75o + 1.25p = 5.25

2b -o + 0p = 0

In [16]:
# Create a system of equations

A = np.matrix([[3,0,1], [.5,.75,1.25], [2, -1, 0]])
B = np.matrix([7,5.25,0])
B = B.T
A_inv = np.linalg.inv(A)
X = A_inv.dot(B)
print(X.T)

[[2. 4. 1.]]


In [23]:
np.matrix.round(A.dot(X), decimals=3)

array([[7.  ],
       [5.25],
       [0.  ]])

In [21]:
np.linalg.solve(A,B)

matrix([[2.],
        [4.],
        [1.]])

# Explain your answer

2 bananas
4 oranges
1 papaya

## Exercise 4

A landlord owns 3 houses, a 1-bedroom, a 2-bedroom , and a 3-bedroom house. 

The total rent he receives is 1240 dollars. 

He needs to make some repairs, where it costs 10% of the 1-bedroom house’s rent for its repairs, 20% of the 2-bedroom for its repairs, and 30% of the 3-bedroom house's rent for its repairs.  The total repair bill for all three houses was $276. 

The 3-bedroom house's rent is twice the 1-bedroom house’s rent. 

How much is the individual rent for three houses?

x = 1br
y = 2br
z = 3br

x + y + z = 1240
.1x + .2y + .3z = 276
-2x + 0y + z = 0

In [24]:
# Create a system of equations
A = np.matrix([[1,1,1], [.1,.2,.3], [-2,0,1]])
B = np.matrix([1240,276,0])
B = B.T
A_inv = np.linalg.inv(A)
X = A_inv.dot(B)
print(X.T)

[[280. 400. 560.]]


In [25]:
np.matrix.round(A.dot(X), decimals=3)

array([[1240.],
       [ 276.],
       [   0.]])

In [26]:
np.linalg.solve(A,B)

matrix([[280.],
        [400.],
        [560.]])

# Explain your answer

The 1br is 280 a month
The 2br is 400 a month
The 3br is 560 a month

## Summary
In this lab, we saw how to use Numpy to solve linear equations by taking inverses and matrix multiplication and also using numpy's `.solve()`. We shall now take these skills forward and see how we can define a simple regression problem using linear algebra and solve it with numpy, only using the skills that we have learnt so far. 